File: src/main/java/me/centralhardware/znatoki/telegram/statistic/eav/types/DateTime.java
package me.centralhardware.znatoki.telegram.statistic.eav.types;

import io.vavr.control.Try;
import io.vavr.control.Validation;
import org.telegram.telegrambots.meta.api.objects.Update;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Optional;

public non-sealed class DateTime implements Type {

    private static final DateTimeFormatter dateFormat    = DateTimeFormatter.ofPattern("dd MM yyyy HH;mm");

    @Override
    public String format(String name, Boolean isOptional) {
        return STR."Введите \{name} в формате dd MM yyyy HH;mm \{isOptional? optionalText: ""}";
    }

    @Override
    public Validation<String, Void> __validate(Update update, String... variants) {
        return validate(update.getMessage().getText());
    }

    @Override
    public Validation<String, Void> validate(String value) {
        return Try.of(() -> LocalDate.parse(value, dateFormat))
                .fold(error -> Validation.invalid("Ошибка обработки даты необходимо ввести в формате: dd MM yyyy"),
                        date -> Validation.valid(null));
    }

    @Override
    public Optional<String> extract(Update update) {
        return Optional.ofNullable(update.getMessage().getText());
    }

    @Override
    public String getName() {
        return "DateTime";
    }
}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/eav/types/DateTime.kt
package me.centralhardware.znatoki.telegram.statistic.eav.types

import arrow.core.Either
import dev.inmo.tgbotapi.extensions.utils.extensions.raw.text
import dev.inmo.tgbotapi.types.message.abstracts.CommonMessage
import dev.inmo.tgbotapi.types.message.content.MessageContent
import me.centralhardware.znatoki.telegram.statistic.eav.types.Type.Companion.OPTIONAL_TEXT
import me.centralhardware.znatoki.telegram.statistic.parseDateTime

object DateTime : Type {

    override fun format(name: String, isOptional: Boolean): String {
        return "Введите $name в формате dd MM yyyy HH;mm ${if (isOptional) OPTIONAL_TEXT else ""}"
    }

    override fun validate(message: CommonMessage<MessageContent>, variants: List<String>): Either<String, Unit> =
        message.text.parseDateTime()
            .fold(
                { Either.Right(Unit) },
                { Either.Left("Ошибка обработки даты необходимо ввести в формате: dd MM yyyy") }
            )

}